### PR TITLE
Fix custom map previews

### LINKED ISSUE #7918: Random crash at startup
I'm suddenly getting crashes after starting the game and then choosing multiplayer.
The snippet is from release-20141029 which crashes immediately when I choose multiplayer.
The exception is identical with playtest-20150405 and playtest-20150408 when I choose "incompatible" in multiplayer mode. It does not crash the playtests if I do not choose to see the "incompatible" servers.

<pre><code>
Platform is Linux
Using SDL 2 with OpenGL renderer
Desktop resolution: 1920x1200
No custom resolution provided, using desktop resolution
Using resolution: 1920x1200
Detected OpenGL version: 3.0
Using OpenAL sound engine
Using default device
Available mods:
        cnc: Tiberian Dawn (release-20141029)
        d2k: Dune 2000 (release-20141029)
        modchooser: Mod Chooser (release-20141029)
        ra: Red Alert (release-20141029)
Loading mod: ra
Exception of type `System.ArgumentException`: A null reference or invalid value was found [GDI+ status: InvalidParameter]
  at System.Drawing.GDIPlus.CheckStatus (Status status) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format, System.Drawing.Imaging.BitmapData bitmapData) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format) [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Drawing.Bitmap:LockBits (System.Drawing.Rectangle,System.Drawing.Imaging.ImageLockMode,System.Drawing.Imaging.PixelFormat)
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: A null reference or invalid value was found [GDI+ status: InvalidParameter]
  at System.Drawing.GDIPlus.CheckStatus (Status status) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format, System.Drawing.Imaging.BitmapData bitmapData) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format) [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Drawing.Bitmap:LockBits (System.Drawing.Rectangle,System.Drawing.Imaging.ImageLockMode,System.Drawing.Imaging.PixelFormat)
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
</code></pre>

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10868670-random-crash-at-startup?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes #7918.

Works for me (Windows/MS runtime) but if somebody could give a mono runtime a whirl that'd be grand.

### CODE REVIEW

**Comment by pchote** [comment]
No joy, but at least we know your new exception works:

```
Exception of type `System.ArgumentException`: src must have a PixelFormat of Format32bppArgb
Parameter name: src
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x000e5] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Util.cs:82 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00010] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SheetBuilder.cs:80 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00141] in /Users/paul/src/OpenRA/OpenRA.Game//Map/MapCache.cs:188 
  at System.Threading.Thread.StartInternal () [0x00016] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/corlib/System.Threading/Thread.cs:696 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: src must have a PixelFormat of Format32bppArgb
Parameter name: src
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x000e5] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Util.cs:82 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00010] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SheetBuilder.cs:80 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00141] in /Users/paul/src/OpenRA/OpenRA.Game//Map/MapCache.cs:188 
  at System.Threading.Thread.StartInternal () [0x00016] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/corlib/System.Threading/Thread.cs:696 
```

**Comment by pchote** [comment]
cc5f5dac06b4efa6a7a976a229708ced72f648a8 fixes the crash but doesn't attempt to recover a correct minimap.  I suspect we want a mix of the two approaches.

**Comment by RoosterDragon** [comment]
Catching all exceptions seems pretty risky - so instead I'm just skipping anything with the wrong pixel format.

**Comment by pchote** [comment]
We're now back to the original crash!
```
Exception of type `System.ArgumentException`: A null reference or invalid value was found [GDI+ status: InvalidParameter]
  at System.Drawing.GDIPlus.CheckStatus (Status status) [0x0009b] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/System.Drawing/System.Drawing/gdipFunctions.cs:212 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format, System.Drawing.Imaging.BitmapData bitmapData) [0x00012] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/System.Drawing/System.Drawing/Bitmap.cs:272 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format) [0x00006] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/System.Drawing/System.Drawing/Bitmap.cs:264 
  at (wrapper remoting-invoke-with-check) System.Drawing.Bitmap:LockBits (System.Drawing.Rectangle,System.Drawing.Imaging.ImageLockMode,System.Drawing.Imaging.PixelFormat)
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x00077] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Util.cs:74 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00010] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SheetBuilder.cs:80 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00141] in /Users/paul/src/OpenRA/OpenRA.Game//Map/MapCache.cs:188 
  at System.Threading.Thread.StartInternal () [0x00016] in /private/tmp/source-mono-mac-4.0.0-branch/bockbuild-mono-4.0.0-branch/profiles/mono-mac-xamarin/build-root/mono-4.0.0/mcs/class/corlib/System.Threading/Thread.cs:696 
```

**Comment by pchote** [comment]
Confirmed that this fixes the crash, but i'm a bit concerned that this might happen again due to some other bitmap weirdness.

**Comment by RoosterDragon** [comment]
Consider me confused then - maybe the pixel format isn't the issue.

A catch-all exception might hurt us by leaving something like the sheet builder in a bad state. Not sure what's best here.

File: OpenRA.Game/Map/MapPreview.cs [hunk 1]
```diff
@@ -155,6 +156,12 @@ public void UpdateRemoteSearch(MapStatus status, MiniYaml yaml)
 						SpawnPoints = spawns;
 
 						CustomPreview = new Bitmap(new MemoryStream(Convert.FromBase64String(r.minimap)));
+						if (CustomPreview.PixelFormat != PixelFormat.Format32bppArgb)
```
Comment (pchote) [inline OpenRA.Game/Map/MapPreview.cs]: If we did the same conversion when loading from disk then we could avoid the discarding.

**Comment by pchote** [comment]
That's a good point that I hadn't considered.  Anyway, the problem I was having was because you're currently only doing the conversion for remote maps while I have the problematic map installed locally.

**Comment by RoosterDragon** [comment]
Derp - amended.

**Comment by obrakmann** [comment]
I still get this exception when the map is installed locally.

```
Exception of type `System.ArgumentException`: src must have a PixelFormat of Format32bppArgb
Parameter name: src
```

**Comment by RoosterDragon** [comment]
I can't see anywhere else a conversion needs to happen, so maybe this is some mono specific weirdness.

I've brought back the check in MapCache to ignore images with the wrong pixel format, so hopefully that prevents the crashes at least.

Somebody on a non-Windows OS might need to pick this up if it's not sufficient, since I can't recreate the issue directly.

**Comment by obrakmann** [comment]
@RoosterDragon: it's not somewhere else, but it seems that the conversion to another PixelFormat doesn't work:

screenshot from 2015-04-11 17 13 12

**Comment by RoosterDragon** [comment]
Huh.

OK, I've added an extension method that will create a blank bitmap in the format we want, then draw the old one into it. Hopefully that actually does the right thing.

**Comment by pchote** [comment]
Confirmed that this fix works :+1:

**Comment by Mailaender** [comment]
Doesn't break the regular server browser minimap retrieval and display. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/4cfb05793a9667326a78284e56ffaf1fde35db0d...36eb5e4fb5ebc7c0ad96357c72594118c54c1f76)

**Pull request merged**